File: mcp/tools/mcp-use/libraries/typescript/packages/agent/src/agents/remote.ts
/** Remote execution support for hosted MCP agents. */

import type { ZodSchema } from "zod";
import { toJSONSchema } from "zod";
import { logger } from "@mcp-use/client";
import type { RunOptions } from "./run_options.js";
import type { BaseMessage } from "./types.js";

// API endpoint constants
const API_CHATS_ENDPOINT = "/api/v1/chats";
const API_CHAT_EXECUTE_ENDPOINT = "/api/v1/chats/{chat_id}/execute";

/**
 * Helper function to normalize run options for remote agent
 */
function normalizeRemoteRunOptions<T>(
  queryOrOptions: string | RunOptions<T>,
  maxSteps?: number,
  manageConnector?: boolean,
  externalHistory?: BaseMessage[],
  outputSchema?: ZodSchema<T>
): {
  query: string;
  maxSteps?: number;
  manageConnector?: boolean;
  externalHistory?: BaseMessage[];
  outputSchema?: ZodSchema<T>;
} {
  // Check if first argument is an options object
  if (typeof queryOrOptions === "object" && queryOrOptions !== null) {
    const options = queryOrOptions as RunOptions<T>;
    return {
      query: options.prompt ?? "",
      maxSteps: options.maxSteps,
      manageConnector: options.manageConnector,
      externalHistory: options.externalHistory,
      outputSchema: options.schema,
    };
  }

  // Old-style positional arguments
  return {
    query: queryOrOptions as string,
    maxSteps,
    manageConnector,
    externalHistory,
    outputSchema,
  };
}

/** Configures a {@link RemoteAgent}. */
export interface RemoteAgentOptions {
  /** Hosted agent identifier. */
  agentId: string;
  /** API key. Defaults to `MCP_USE_API_KEY`. */
  apiKey?: string;
  /** API origin. Defaults to `https://cloud.manufact.com`. */
  baseUrl?: string;
}

/** Executes a hosted MCP agent through the mcp-use remote API. */
export class RemoteAgent {
  private agentId: string;
  private apiKey: string;
  private baseUrl: string;
  private chatId: string | null = null;

  /**
   * @param options - Hosted agent identifier and API connection settings.
   * @throws Error if no API key is supplied or available from
   * `MCP_USE_API_KEY`.
   */
  constructor(options: RemoteAgentOptions) {
    this.agentId = options.agentId;
    this.baseUrl = options.baseUrl ?? "https://cloud.manufact.com";

    // Handle API key validation
    const apiKey =
      options.apiKey ??
      (typeof process !== "undefined" && process.env?.MCP_USE_API_KEY);
    if (!apiKey) {
      throw new Error(
        "API key is required for remote execution. " +
          "Please provide it as a parameter or set the MCP_USE_API_KEY environment variable. " +
          "You can get an API key from https://cloud.manufact.com"
      );
    }
    this.apiKey = apiKey;
  }

  private pydanticToJsonSchema<T>(schema: ZodSchema<T>): any {
    /**
     * Convert a Zod schema to JSON schema for API transmission.
     */
    return toJSONSchema(schema);
  }

  private parseStructuredResponse<T>(
    responseData: any,
    outputSchema: ZodSchema<T>
  ): T {
    /**
     * Parse the API response into the structured output format.
     */
    let resultData: any;

    // Handle different response formats
    if (typeof responseData === "object" && responseData !== null) {
      if ("result" in responseData) {
        const outerResult = responseData.result;
        // Check if this is a nested result structure (agent execution response)
        if (
          typeof outerResult === "object" &&
          outerResult !== null &&
          "result" in outerResult
        ) {
          // Extract the actual structured output from the nested result
          resultData = outerResult.result;
        } else {
          // Use the outer result directly
          resultData = outerResult;
        }
      } else {
        resultData = responseData;
      }
    } else if (typeof responseData === "string") {
      try {
        resultData = JSON.parse(responseData);
      } catch {
        // If it's not valid JSON, try to create the model from the string content
        resultData = { content: responseData };
      }
    } else {
      resultData = responseData;
    }

    // Parse into the Zod schema
    try {
      return outputSchema.parse(resultData);
    } catch (e) {
      logger.warn(`Failed to parse structured output: ${e}`);
      // Fallback: try to parse it as raw content if the schema has a content field
      const schemaShape = (outputSchema as any)._def?.shape();
      if (schemaShape && "content" in schemaShape) {
        return outputSchema.parse({ content: String(resultData) });
      }
      throw e;
    }
  }

  private async createChatSession(): Promise<string> {
    /**
     * Create a persistent chat session for the agent.
     */
    const chatPayload = {
      title: `Remote Agent Session - ${this.agentId}`,
      agent_id: this.agentId,
      type: "agent_execution",
    };

    const headers = {
      "Content-Type": "application/json",
      "x-api-key": this.apiKey,
    };
    const chatUrl = `${this.baseUrl}${API_CHATS_ENDPOINT}`;

    logger.debug(`📝 Creating chat session for agent ${this.agentId}`);

    try {
      const response = await fetch(chatUrl, {
        method: "POST",
        headers,
        body: JSON.stringify(chatPayload),
      });

      if (!response.ok) {
        const responseText = await response.text();
        const statusCode = response.status;

        if (statusCode === 404) {
          throw new Error(
            `Agent not found: Agent '${this.agentId}' does not exist or you don't have access to it. ` +
              "Please verify the agent ID and ensure it exists in your account."
          );
        }
        throw new Error(
          `Failed to create chat session: ${statusCode} - ${responseText}`
        );
      }

      const chatData = await response.json();
      const chatId = chatData.id;
      logger.debug(`✅ Chat session created: ${chatId}`);
      return chatId;
    } catch (e) {
      if (e instanceof Error) {
        throw new TypeError(`Failed to create chat session: ${e.message}`);
      }
      throw new Error(`Failed to create chat session: ${String(e)}`);
    }
  }

  /**
   * Runs the remote agent and returns its final text.
   *
   * @param options - Input and per-run execution settings.
   * @returns Final agent text.
   */
  public async run(options: RunOptions): Promise<string>;

  /**
   * Runs the remote agent and parses its result with `options.schema`.
   *
   * @param options - Input, schema, and per-run execution settings.
   * @returns The schema-validated result.
   */
  public async run<T>(options: RunOptions<T>): Promise<T>;

  /**
   * Runs the remote agent and returns a promise for the final result.
   * @deprecated Use the options object instead: `run({ prompt, maxSteps, ... })`.
   */
  public async run<T = string>(
    query: string,
    maxSteps?: number,
    manageConnector?: boolean,
    externalHistory?: BaseMessage[],
    outputSchema?: ZodSchema<T>
  ): Promise<T>;

  public async run<T = string>(
    queryOrOptions: string | RunOptions<T>,
    maxSteps?: number,
    manageConnector?: boolean,
    externalHistory?: BaseMessage[],
    outputSchema?: ZodSchema<T>
  ): Promise<T> {
    /**
     * Run a query on the remote agent.
     */
    // Normalize input to internal parameters
    const {
      query,
      maxSteps: steps,
      externalHistory: history,
      outputSchema: schema,
    } = normalizeRemoteRunOptions(
      queryOrOptions,
      maxSteps,
      manageConnector,
      externalHistory,
      outputSchema
    );

    if (history !== undefined) {
      logger.warn("External history is not yet supported for remote execution");
    }

    try {
      logger.debug(`🌐 Executing query on remote agent ${this.agentId}`);

      // Step 1: Create a chat session for this agent (only if we don't have one)
      if (this.chatId === null) {
        this.chatId = await this.createChatSession();
      }

      const chatId = this.chatId;

      // Step 2: Execute the agent within the chat context
      const executionPayload: any = {
        query,
        max_steps: steps ?? 10,
      };

      // Add structured output schema if provided
      if (schema) {
        executionPayload.output_schema = this.pydanticToJsonSchema(schema);
        logger.debug(`🔧 Using structured output with schema`);
      }

      const headers = {
        "Content-Type": "application/json",
        "x-api-key": this.apiKey,
      };
      const executionUrl = `${this.baseUrl}${API_CHAT_EXECUTE_ENDPOINT.replace("{chat_id}", chatId)}`;
      logger.debug(`🚀 Executing agent in chat ${chatId}`);

      const response = await fetch(executionUrl, {
        method: "POST",
        headers,
        body: JSON.stringify(executionPayload),
        signal: AbortSignal.timeout(300000), // 5 minute timeout
      });

      if (!response.ok) {
        const responseText = await response.text();
        const statusCode = response.status;

        // Provide specific error messages based on status code
        if (statusCode === 401) {
          logger.error(`❌ Authentication failed: ${responseText}`);
          throw new Error(
            "Authentication failed: Invalid or missing API key. " +
              "Please check your API key and ensure the MCP_USE_API_KEY environment variable is set correctly."
          );
        } else if (statusCode === 403) {
          logger.error(`❌ Access forbidden: ${responseText}`);
          throw new Error(
            `Access denied: You don't have permission to execute agent '${this.agentId}'. ` +
              "Check if the agent exists and you have the necessary permissions."
          );
        } else if (statusCode === 404) {
          logger.error(`❌ Agent not found: ${responseText}`);
          throw new Error(
            `Agent not found: Agent '${this.agentId}' does not exist or you don't have access to it. ` +
              "Please verify the agent ID and ensure it exists in your account."
          );
        } else if (statusCode === 422) {
          logger.error(`❌ Validation error: ${responseText}`);
          throw new Error(
            `Request validation failed: ${responseText}. ` +
              "Please check your query parameters and output schema format."
          );
        } else if (statusCode === 500) {
          logger.error(`❌ Server error: ${responseText}`);
          throw new Error(
            "Internal server error occurred during agent execution. " +
              "Please try again later or contact support if the issue persists."
          );
        } else {
          logger.error(
            `❌ Remote execution failed with status ${statusCode}: ${responseText}`
          );
          throw new Error(
            `Remote agent execution failed: ${statusCode} - ${responseText}`
          );
        }
      }

      const result = await response.json();
      logger.debug(`🔧 Response: ${JSON.stringify(result)}`);
      logger.debug("✅ Remote execution completed successfully");

      // Check for error responses (even with 200 status)
      if (typeof result === "object" && result !== null) {
        // Check for actual error conditions (not just presence of error field)
        if (result.status === "error" || result.error !== null) {
          const errorMsg = result.error ?? String(result);
          logger.error(`❌ Remote agent execution failed: ${errorMsg}`);
          throw new Error(`Remote agent execution failed: ${errorMsg}`);
        }

        // Check if the response indicates agent initialization failure
        if (String(result).includes("failed to initialize")) {
          logger.error(`❌ Agent initialization failed: ${result}`);
          throw new Error(
            "Agent initialization failed on remote server. " +
              "This usually indicates:\n" +
              "• Invalid agent configuration (LLM model, system prompt)\n" +
              "• Missing or invalid MCP server configurations\n" +
              "• Network connectivity issues with MCP servers\n" +
              "• Missing environment variables or credentials\n" +
              `Raw error: ${result}`
          );
        }
      }

      // Handle structured output
      if (schema) {
        return this.parseStructuredResponse(result, schema);
      }

      // Regular string output
      if (typeof result === "object" && result !== null && "result" in result) {
        return result.result as T;
      } else if (typeof result === "string") {
        return result as T;
      } else {
        return String(result) as T;
      }
    } catch (e) {
      if (e instanceof Error) {
        // Check for specific error types
        if (e.name === "AbortError") {
          logger.error(`❌ Remote execution timed out: ${e}`);
          throw new Error(
            "Remote agent execution timed out. The server may be overloaded or the query is taking too long to " +
              "process. Try again or use a simpler query."
          );
        }
        logger.error(`❌ Remote execution error: ${e}`);
        throw new Error(`Remote agent execution failed: ${e.message}`);
      }
      logger.error(`❌ Remote execution error: ${e}`);
      throw new Error(`Remote agent execution failed: ${String(e)}`);
    }
  }

  /**
   * Runs the remote agent through an async-generator interface.
   *
   * The current remote API does not emit intermediate values. Read the
   * generator's return value for the final result.
   *
   * @param options - Input and per-run execution settings.
   * @returns An async generator whose return value is the final text.
   */
  public stream(options: RunOptions): AsyncGenerator<any, string, void>;

  /**
   * Runs structured remote execution through an async-generator interface.
   *
   * @param options - Input, schema, and per-run execution settings.
   * @returns An async generator whose return value is schema validated.
   */
  public stream<T>(options: RunOptions<T>): AsyncGenerator<any, T, void>;

  /**
   * Streams the remote agent execution.
   * @deprecated Use the options object instead: `stream({ prompt, maxSteps, ... })`.
   */
  public stream<T = string>(
    query: string,
    maxSteps?: number,
    manageConnector?: boolean,
    externalHistory?: BaseMessage[],
    outputSchema?: ZodSchema<T>
  ): AsyncGenerator<any, T, void>;

  // eslint-disable-next-line require-yield
  public async *stream<T = string>(
    queryOrOptions: string | RunOptions<T>,
    maxSteps?: number,
    manageConnector?: boolean,
    externalHistory?: BaseMessage[],
    outputSchema?: ZodSchema<T>
  ): AsyncGenerator<any, T, void> {
    /**
     * Stream implementation for remote agent - currently just wraps run.
     * In the future, this could be enhanced to support actual streaming from the API.
     */
    const result = await this.run(
      queryOrOptions as any,
      maxSteps,
      manageConnector,
      externalHistory,
      outputSchema
    );
    return result;
  }

  /** Releases local remote-agent state. */
  public async close(): Promise<void> {
    /**
     * Close the remote agent connection.
     */
    logger.debug("🔌 Remote agent client closed");
    // In the future, we might want to delete the chat session here
    // if (this.chatId) {
    //   await this.deleteChatSession(this.chatId)
    // }
  }
}

File: mcp/tools/mcp-use/libraries/python/mcp_use/agents/remote.py
"""
Remote agent implementation for executing agents via API.
"""

import json
import os
from collections.abc import AsyncGenerator
from typing import Any, TypeVar
from uuid import UUID

import httpx
from langchain_core.messages import BaseMessage
from pydantic import BaseModel

from mcp_use.logging import logger

T = TypeVar("T", bound=BaseModel)

# API endpoint constants
API_CHATS_ENDPOINT = "/api/v1/chats/get-or-create"
API_CHAT_STREAM_ENDPOINT = "/api/v1/chats/{chat_id}/stream"
API_CHAT_DELETE_ENDPOINT = "/api/v1/chats/{chat_id}"

UUID_ERROR_MESSAGE = """A UUID is a 36 character string of the format xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx \n
Example: 123e4567-e89b-12d3-a456-426614174000
To generate a UUID, you can use the following command:
import uuid

# Generate a random UUID
my_uuid = uuid.uuid4()
print(my_uuid)
"""


class RemoteAgent:
    """Agent that executes remotely via API."""

    def __init__(
        self,
        agent_id: str,
        chat_id: str | None = None,
        api_key: str | None = None,
        base_url: str = "https://cloud.manufact.com",
    ):
        """Initialize remote agent.

        Args:
            agent_id: The ID of the remote agent to execute
            chat_id: The ID of the chat session to use. If None, a new chat session will be created.
            api_key: API key for authentication. If None, will check MCP_USE_API_KEY env var
            base_url: Base URL for the remote API
        """

        if chat_id is not None:
            try:
                chat_id = str(UUID(chat_id))
            except ValueError as e:
                raise ValueError(
                    f"Invalid chat ID: {chat_id}, make sure to provide a valid UUID.\n{UUID_ERROR_MESSAGE}"
                ) from e

        self.agent_id = agent_id
        self.chat_id = chat_id
        self._session_established = False
        self.base_url = base_url

        # Handle API key validation
        if api_key is None:
            api_key = os.getenv("MCP_USE_API_KEY")
        if not api_key:
            raise ValueError(
                "API key is required for remote execution. "
                "Please provide it as a parameter or set the MCP_USE_API_KEY environment variable. "
                "You can get an API key from https://cloud.manufact.com"
            )

        self.api_key = api_key
        # Configure client with reasonable timeouts for agent execution
        self._client = httpx.AsyncClient(
            timeout=httpx.Timeout(
                connect=10.0,  # 10 seconds to establish connection
                read=300.0,  # 5 minutes to read response (agents can take time)
                write=10.0,  # 10 seconds to send request
                pool=10.0,  # 10 seconds to get connection from pool
            )
        )

    def _pydantic_to_json_schema(self, model_class: type[T]) -> dict[str, Any]:
        """Convert a Pydantic model to JSON schema for API transmission.

        Args:
            model_class: The Pydantic model class to convert

        Returns:
            JSON schema representation of the model
        """
        return model_class.model_json_schema()

    def _parse_structured_response(self, response_data: Any, output_schema: type[T]) -> T:
        """Parse the API response into the structured output format.

        Args:
            response_data: Raw response data from the API
            output_schema: The Pydantic model to parse into

        Returns:
            Parsed structured output
        """
        # Handle different response formats
        if isinstance(response_data, dict):
            if "result" in response_data:
                outer_result = response_data["result"]
                # Check if this is a nested result structure (agent execution response)
                if isinstance(outer_result, dict) and "result" in outer_result:
                    # Extract the actual structured output from the nested result
                    result_data = outer_result["result"]
                else:
                    # Use the outer result directly
                    result_data = outer_result
            else:
                result_data = response_data
        elif isinstance(response_data, str):
            try:
                result_data = json.loads(response_data)
            except json.JSONDecodeError:
                # If it's not valid JSON, try to create the model from the string content
                result_data = {"content": response_data}
        else:
            result_data = response_data

        # Parse into the Pydantic model
        try:
            logger.info(f"🔍 Attempting to validate result_data against {output_schema.__name__}")
            logger.info(f"🔍 Result data type: {type(result_data)}")
            logger.info(f"🔍 Result data: {result_data}")
            return output_schema.model_validate(result_data)
        except Exception as e:
            logger.warning(f"❌ Failed to parse structured output: {e}")
            logger.warning(f"🔍 Validation error details: {type(e).__name__}: {str(e)}")
            logger.warning(f"🔍 Result data that failed validation: {result_data}")

            # Fallback: try to parse it as raw content if the model has a content field
            if hasattr(output_schema, "model_fields") and "content" in output_schema.model_fields:
                logger.info("🔄 Attempting fallback with content field")
                try:
                    fallback_result = output_schema.model_validate({"content": str(result_data)})
                    logger.info("✅ Fallback parsing succeeded")
                    return fallback_result
                except Exception as fallback_e:
                    logger.error(f"❌ Fallback parsing also failed: {fallback_e}")
                    raise
            raise

    async def _upsert_chat_session(self) -> str:
        """Create or resume a persistent chat session for the agent via upsert.

        Returns:
            The chat session ID
        """
        chat_payload = {
            "id": self.chat_id,  # Include chat_id for resuming or None for creating
            "title": f"Remote Agent Session - {self.agent_id}",
            "agent_id": self.agent_id,
            "type": "agent_execution",
        }

        headers = {"Content-Type": "application/json", "x-api-key": self.api_key}
        chat_url = f"{self.base_url}{API_CHATS_ENDPOINT}"

        logger.info(f"📝 [{self.chat_id}] Upserting chat session for agent {self.agent_id}")

        try:
            chat_response = await self._client.post(chat_url, json=chat_payload, headers=headers)
            chat_response.raise_for_status()

            chat_data = chat_response.json()
            chat_id = chat_data["id"]
            if chat_response.status_code == 201:
                logger.info(f"✅ [{self.chat_id}] New chat session created")
            else:
                logger.info(f"✅ [{self.chat_id}] Resumed chat session")

            return chat_id

        except httpx.HTTPStatusError as e:
            status_code = e.response.status_code
            response_text = e.response.text

            if status_code == 404:
                raise RuntimeError(
                    f"Agent not found: Agent '{self.agent_id}' does not exist or you don't have access to it. "
                    "Please verify the agent ID and ensure it exists in your account."
                ) from e
            else:
                raise RuntimeError(f"Failed to create chat session: {status_code} - {response_text}") from e
        except Exception as e:
            raise RuntimeError(f"Failed to create chat session: {str(e)}") from e

    async def stream(
        self,
        query: str,
        max_steps: int | None = None,
        external_history: list[BaseMessage] | None = None,
        output_schema: type[T] | None = None,
    ) -> AsyncGenerator[str, None]:
        """Stream the execution of a query on the remote agent using HTTP streaming."""
        if external_history is not None:
            logger.warning("External history is not yet supported for remote execution")

        if not self._session_established:
            logger.info(f"🔧 [{self.chat_id}] Establishing chat session for agent {self.agent_id}")
            self.chat_id = await self._upsert_chat_session()
            self._session_established = True

        chat_id = self.chat_id
        stream_url = f"{self.base_url}{API_CHAT_STREAM_ENDPOINT.format(chat_id=chat_id)}"

        # Prepare the request payload
        request_payload = {"messages": [{"role": "user", "content": query}], "max_steps": max_steps or 30}
        if output_schema is not None:
            request_payload["output_schema"] = self._pydantic_to_json_schema(output_schema)

        headers = {"Content-Type": "application/json", "x-api-key": self.api_key, "Accept": "text/event-stream"}

        try:
            logger.info(f"🌐 [{self.chat_id}] Connecting to HTTP stream for agent {self.agent_id}")

            async with self._client.stream("POST", stream_url, headers=headers, json=request_payload) as response:
                logger.info(f"✅ [{self.chat_id}] HTTP stream connection established")

                if response.status_code != 200:
                    error_text = await response.aread()
                    raise RuntimeError(f"Failed to stream from remote agent: {error_text.decode()}")

                # Read the streaming response line by line
                try:
                    async for line in response.aiter_lines():
                        if line:
                            yield line
                except UnicodeDecodeError as e:
                    logger.error(f"❌ [{self.chat_id}] UTF-8 decoding error at position {e.start}: {e.reason}")
                    logger.error(f"❌ [{self.chat_id}] Error occurred while reading stream for agent {self.agent_id}")
                    # Try to read raw bytes and decode with error handling
                    logger.info(f"🔄 [{self.chat_id}] Attempting to read raw bytes with error handling...")
                logger.info(f"✅ [{self.chat_id}] Agent execution stream completed")

        except httpx.HTTPStatusError as e:
            status_code = e.response.status_code
            response_text = e.response.text

            if status_code == 404:
                raise RuntimeError(f"Chat or agent not found: {response_text}") from e
            else:
                raise RuntimeError(f"Failed to stream from remote agent: {status_code} - {response_text}") from e
        except Exception as e:
            logger.error(f"❌ [{self.chat_id}] An error occurred during HTTP streaming: {e}")
            raise RuntimeError(f"Failed to stream from remote agent: {str(e)}") from e

    async def run(
        self,
        query: str,
        max_steps: int | None = None,
        external_history: list[BaseMessage] | None = None,
        output_schema: type[T] | None = None,
    ) -> str | T:
        """
        Executes the agent and returns the final result.
        This method uses HTTP streaming to avoid timeouts for long-running tasks.
        It consumes the entire stream and returns only the final result.
        """
        final_result = None
        steps_taken = 0
        finished = False

        try:
            # Consume the ENTIRE stream to ensure proper execution
            async for event in self.stream(query, max_steps, external_history, output_schema):
                logger.debug(f"[{self.chat_id}] Processing stream event: {event}...")

                # Parse AI SDK format events to extract final result
                # The events follow the AI SDK streaming protocol
                if event.startswith("0:"):  # Text event
                    try:
                        text_data = json.loads(event[2:])  # Remove "0:" prefix
                        # Normal text accumulation
                        if final_result is None:
                            final_result = ""
                        final_result += text_data
                        result_preview = final_result[:200] if len(final_result) > 200 else final_result
                        logger.debug(f"Accumulated text result: {result_preview}...")
                    except json.JSONDecodeError:
                        logger.warning(f"Failed to parse text event: {event[:100]}")
                        continue

                elif event.startswith("9:"):  # Tool call event
                    steps_taken += 1
                    logger.debug(f"Tool call executed, total steps: {steps_taken}")

                elif event.startswith("d:"):  # Finish event
                    logger.debug("Received finish event, marking as finished")
                    finished = True
                    # Continue consuming to ensure stream cleanup

                elif event.startswith("3:"):  # Error event
                    try:
                        error_data = json.loads(event[2:])
                        error_msg = error_data if isinstance(error_data, str) else json.dumps(error_data)
                        raise RuntimeError(f"Agent execution failed: {error_msg}")
                    except json.JSONDecodeError as e:
                        raise RuntimeError("Agent execution failed with unknown error") from e

                elif event.startswith("f:"):  # Structured final event
                    try:
                        structured_data = json.loads(event[2:])  # Remove "f:" prefix
                        logger.info(f"📋 [{self.chat_id}] Received structured final event")

                        # Replace accumulated text with structured output
                        final_result = structured_data
                        logger.info(f"📋 [{self.chat_id}] Replaced accumulated text with structured output")
                    except json.JSONDecodeError:
                        logger.warning(f"Failed to parse structured final event: {event[:100]}")
                        continue

            # Log completion of stream consumption
            logger.info(f"Stream consumption complete. Finished: {finished}, Steps taken: {steps_taken}")

            if final_result is None:
                logger.warning(f"No final result captured from stream (structured output: {output_schema is not None})")
                final_result = ""  # Return empty string instead of error message

            # For structured output, try to parse the result
            if output_schema:
                logger.info(f"🔍 Attempting structured output parsing for schema: {output_schema.__name__}")
                logger.info(f"🔍 Raw final result type: {type(final_result)}")
                logger.info(f"🔍 Raw final result length: {len(str(final_result)) if final_result else 0}")
                logger.info(f"🔍 Raw final result preview: {str(final_result)[:500] if final_result else 'None'}...")

                if isinstance(final_result, str) and final_result:
                    try:
                        # Try to parse as JSON first
                        parsed_result = json.loads(final_result)
                        logger.info("✅ Successfully parsed structured result as JSON")
                        return self._parse_structured_response(parsed_result, output_schema)
                    except json.JSONDecodeError as e:
                        logger.warning(f"❌ Could not parse result as JSON: {e}")
                        logger.warning(f"🔍 Raw string content: {final_result[:1000]}...")
                        # Try to parse directly
                        return self._parse_structured_response({"content": final_result}, output_schema)
                else:
                    logger.warning(f"❌ Final result is empty or not string: {final_result}")
                    # Try to parse the result directly
                    return self._parse_structured_response(final_result, output_schema)

            # Regular string output
            return final_result if isinstance(final_result, str) else str(final_result)

        except RuntimeError:
            raise
        except Exception as e:
            logger.error(f"Error executing agent: {e}")
            raise RuntimeError(f"Failed to execute agent: {str(e)}") from e

    async def close(self) -> None:
        """Close the HTTP client."""
        await self._client.aclose()
        logger.info("🔌 Remote agent client closed")
